contracts/Governance/Timelock.sol

```diff
@@ -1,123 +1,177 @@
+    function cancelTransactionsBatch(bytes32 txParamsHash) external onlyOwner {
+        require(queuedTransactions[txParamsHash] != TransactionStatus.NonExistent, "Timelock: transaction is not queued");
 
-        bytes32 txHash = keccak256(abi.encode(target, value, signature, data, eta));
-        require(queuedTransactions[txHash], "Timelock::executeTransaction: Transaction hasn't been queued.");
-        require(getBlockTimestamp() >= eta, "Timelock::executeTransaction: Transaction hasn't surpassed time lock.");
-        require(getBlockTimestamp() <= eta + GRACE_PERIOD, "Timelock::executeTransaction: Transaction is stale.");
+        delete queuedTransactions[txParamsHash];
 
-        queuedTransactions[txHash] = false;
+        emit CancelledTransactionsBatch(txParamsHash);
+    }
 
-        bytes memory callData;
+    function approveTransactionsBatch(bytes32 txParamsHash) external onlyOwner {
+        _approveTransactionsBatchInternal(txParamsHash);
+    }
 
-        if (bytes(signature).length == 0) {
-            callData = data;
-        } else {
-            callData = abi.encodePacked(bytes4(keccak256(bytes(signature))), data);
-        }
+    function executeTransactionsBatch(Transaction[] memory transactions, uint256 executionStartTimestamp) external payable onlyOwner {
+        _executeTransactionsBatchInternal(transactions, executionStartTimestamp);
+    }
+
+    function approveAndExecuteTransactionsBatch(Transaction[] memory transactions, uint256 executionStartTimestamp) external payable onlyOwner {
+        bytes32 txParamsHash = keccak256(abi.encode(transactions, executionStartTimestamp));
+
+        _approveTransactionsBatchInternal(txParamsHash);
+        _executeTransactionsBatchInternal(transactions, executionStartTimestamp);
+    }
+
+    function approveAndExecuteTransactionsBatchRaw(bytes calldata txParamsData) external payable onlyOwner {
+        bytes32 txParamsHash = keccak256(txParamsData);
+        (Transaction[] memory transactions, uint256 executionStartTimestamp) = abi.decode(txParamsData, (Transaction[], uint256));
+
+        _approveTransactionsBatchInternal(txParamsHash);
+        _executeTransactionsBatchInternal(transactions, executionStartTimestamp);
+    }
+
+    function _approveTransactionsBatchInternal(bytes32 txParamsHash) internal onlyOwner {
+        require(queuedTransactions[txParamsHash] == TransactionStatus.Queued, "Timelock: transaction is not queued");
+
+        queuedTransactions[txParamsHash] = TransactionStatus.Approved;
 
-        // Execute the call
-        (bool success, bytes memory returnData) = target.call{value: value}(callData);
-        require(success, "Timelock::executeTransaction: Transaction execution reverted.");
+        emit ApprovedTransactionsBatch(txParamsHash);
+    }
+
+    function _executeTransactionsBatchInternal(Transaction[] memory transactions, uint256 executionStartTimestamp) internal {
```

Comment:
Thinking out loud here - do we need a `nonReentrant` guard here?
We're executing a transaction so going to another contract here..

@tkeidar FYI

Reply:
I don't think so, there's nothing to steal from timelock, plus we delete the queued transaction before we execute it

Reply:
Thinking out loud here, can't a malicious contract reenter this contract with a different transaction data and execute it?

Reply:
![image](https://user-images.githubusercontent.com/79972464/168073671-e936a31a-84e9-4526-a34b-7c2418ed70e3.png)

Nope, we're good here. Once (2) executes, subsequent (1) execution will fail.

Reply:
But on number 2 we're deleting the transaction with specific params hash.. What I'm saying is that someone can come with a different set of params we haven't deleted yet

Maybe worth talking about it online to finalize it quicker :)

Reply:
IMO there is no risk in adding it and it will cover any potential risk we're not thinking about here..
WDYS about adding it?